File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/25-SingleSourceShortestPath/LeetCode/LeetCode-0743-2.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        /// This time we run the Bellmann-Ford algorithm
        int maxCost = INT_MAX/2;
        vector<int> cost(n+1, maxCost);
        cost[k] = 0;
        
        for (int iter=0; iter<n-1; iter++){
            for (int i=0; i<times.size(); i++){
                int u = times[i][0];
                int v = times[i][1];
                int w = times[i][2];
                
                cost[v] = min(cost[v], cost[u]+w); // Relax(u, v)
            } // end-for         
        } // end-for
        
        int maxVal = *std::max_element(cost.begin()+1, cost.end());        
        return maxVal == maxCost? -1: maxVal;
    } // end-for
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/25-SingleSourceShortestPath/LeetCode/LeetCode-0743-2.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        """
        This time we run the Bellmann-Ford algorithm
        """        
        maxDist = 100*100
        cost = [maxDist]*(n+1)
        cost[0] = 0      # There is no node with id=0. So, make its distance 0
        cost[k] = 0
        
        for ite in range(n-1):
            for u, v, w in times:
                cost[v] = min(cost[v], cost[u]+w) # Relax(u, v)
        
        maxVal = max(cost)
        return -1 if maxVal == maxDist else maxVal
